official/transformer/transformer_main.py

```diff
@@ -245,145 +267,198 @@ def train_schedule(
+      help=flags_core.help_wrap(
+          "Path to source file containing text translate when calculating the "
+          "official BLEU score. --bleu_source, --bleu_ref, and --vocab_file "
+          "must be set. Use the flag --stop_threshold to stop the script based "
+          "on the uncased BLEU score."))
+  flags.DEFINE_string(
+      name="vocab_file", short_name="vf", default=VOCAB_FILE,
+      help=flags_core.help_wrap(
+          "Name of vocabulary file containing subtokens for subtokenizing the "
+          "bleu_source file. This file is expected to be in the directory "
+          "defined by --data_dir."))
+
+  flags_core.set_defaults(data_dir="/tmp/translate_ende",
+                          model_dir="/tmp/transformer_model",
+                          batch_size=None,
+                          train_epochs=None)
+
+  @flags.multi_flags_validator(
+      ["train_epochs", "train_steps"],
+      message="Both --train_steps and --train_epochs were set. Only one may be "
+              "defined.")
+  def _check_train_limits(flag_dict):
+    return flag_dict["train_epochs"] is None or flag_dict["train_steps"] is None
+
+  @flags.multi_flags_validator(
+      ["data_dir", "bleu_source", "bleu_ref", "vocab_file"],
+      message="--bleu_source, --bleu_ref, and/or --vocab_file don't exist. "
+              "Please ensure that the file paths are correct.")
+  def _check_bleu_files(flags_dict):
+    """Validate files when bleu_source and bleu_ref are defined."""
+    if flags_dict["bleu_source"] is None or flags_dict["bleu_ref"] is None:
+      return True
+    # Ensure that bleu_source, bleu_ref, and vocab files exist.
+    vocab_file_path = os.path.join(
+        flags_dict["data_dir"], flags_dict["vocab_file"])
+    return all([
+        tf.gfile.Exists(flags_dict["bleu_source"]),
+        tf.gfile.Exists(flags_dict["bleu_ref"]),
+        tf.gfile.Exists(vocab_file_path)])
+
+
+def run_transformer(flags_obj):
+  """Create tf.Estimator to train and evaluate transformer model.
+
+  Args:
+    flags_obj: Object containing parsed flag values.
+  """
   # Determine training schedule based on flags.
-  if FLAGS.train_steps is not None and FLAGS.train_epochs is not None:
-    raise ValueError("Both --train_steps and --train_epochs were set. Only one "
-                     "may be defined.")
-  if FLAGS.train_steps is not None:
-    train_eval_iterations = FLAGS.train_steps // FLAGS.steps_between_eval
-    single_iteration_train_steps = FLAGS.steps_between_eval
+  if flags_obj.train_steps is not None:
+    train_eval_iterations = (
+        flags_obj.train_steps // flags_obj.steps_between_evals)
+    single_iteration_train_steps = flags_obj.steps_between_evals
     single_iteration_train_epochs = None
   else:
-    if FLAGS.train_epochs is None:
-      FLAGS.train_epochs = DEFAULT_TRAIN_EPOCHS
-    train_eval_iterations = FLAGS.train_epochs // FLAGS.epochs_between_eval
+    if flags_obj.train_epochs is None:
```

Comment:
We have adopted a convention of not modifying flags objects, and instead calling getter functions to retrieve values. See `official.utils.flags._performance.get_loss_scale()` as an example.

Reply:
Yes, it makes sense not to alter the value of the flags after. It would great if these conventions can be listed (maybe as a sub-checkbox in karmel's list above).

Reply:
Would you mind adding an "immutability" section to flags/README.md?